### ISSUE TITLE
ProcessCanceledException while processing MicroProfile diagnostics

### ISSUE THREAD

**Issue by turkeylurkey** [OpenLiberty/liberty-tools-intellij#612]
This exception stops correct processing of diagnostics. 
```
WARN - io.openliberty.tools.intellij.lsp4mp4ij.psi.internal.core.java.corrections.JavaDiagnosticsDefinition - Error while calling collectDiagnostics
com.intellij.openapi.progress.ProcessCanceledException
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorBase.throwIfCanceled(AbstractProgressIndicatorBase.java:158)
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorBase.checkCanceled(AbstractProgressIndicatorBase.java:148)
	at com.intellij.openapi.progress.util.ProgressWrapper.checkCanceledRaw(ProgressWrapper.java:67)
	at com.intellij.openapi.progress.util.ProgressWrapper.checkCanceled(ProgressWrapper.java:74)
	at com.intellij.openapi.progress.impl.CoreProgressManager.doCheckCanceled(CoreProgressManager.java:141)
	at com.intellij.openapi.progress.ProgressManager.checkCanceled(ProgressManager.java:231)
	at com.intellij.openapi.progress.ProgressIndicatorProvider.checkCanceled(ProgressIndicatorProvider.java:23)
	at com.intellij.psi.impl.source.tree.CompositeElement.getPsi(CompositeElement.java:676)
	at com.intellij.psi.impl.source.tree.CompositeElement.getChildrenAsPsiElements(CompositeElement.java:377)
	at com.intellij.psi.impl.source.PsiFileImpl.getChildren(PsiFileImpl.java:732)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.internal.metrics.java.MicroProfileMetricsDiagnosticsParticipant.collectDiagnostics(MicroProfileMetricsDiagnosticsParticipant.java:63)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.internal.core.java.corrections.JavaDiagnosticsDefinition.collectDiagnostics(JavaDiagnosticsDefinition.java:70)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.core.java.diagnostics.DiagnosticsHandler.lambda$collectDiagnostics$5(DiagnosticsHandler.java:86)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.core.java.diagnostics.DiagnosticsHandler.lambda$collectDiagnostics$7(DiagnosticsHandler.java:85)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.kt:105)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.core.java.diagnostics.DiagnosticsHandler.collectDiagnostics(DiagnosticsHandler.java:72)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.core.java.diagnostics.DiagnosticsHandler.collectDiagnostics(DiagnosticsHandler.java:58)
	at io.openliberty.tools.intellij.lsp4mp4ij.psi.core.PropertiesManagerForJava.diagnostics(PropertiesManagerForJava.java:272)
	at io.openliberty.tools.intellij.lsp4mp.lsp.MicroProfileLanguageClient.lambda$getJavaDiagnostics$7(MicroProfileLanguageClient.java:196)
	at org.microshed.lsp4ij.internal.PromiseToCompletableFuture.lambda$nonBlockingReadActionPromise$2(PromiseToCompletableFuture.java:137)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$OTelMonitor.callWrapped(NonBlockingReadActionImpl.java:832)
```

**Comment by angelozerr** [comment]
As `com.intellij.openapi.progress.ProcessCanceledException` can occur sometimes (a usecase is when Java Psi is not ready I think), I have introduced `PromiseToCompletableFuture` and the idea is to let's delegate the cancel to PromiseToCompletableFuture which will cancel the original CompletableFuture of the LS.

In other words to fix this problem, you have to avoid catching all exception like:
https://github.com/OpenLiberty/liberty-tools-intellij/blob/7b6689af4c8f9afd2e36ca3c05798c0362e702c7/src/main/java/io/openliberty/tools/intellij/lsp4mp4ij/psi/internal/core/java/corrections/JavaDiagnosticsDefinition.java#L72

See sample at https://github.com/redhat-developer/intellij-quarkus/blob/abc6df95d1000855a6e918576e12ad1430a76213/src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/internal/jaxrs/java/JaxRsCodeLensParticipant.java#L101

IndexNotReadyException must b erethrown because  `PromiseToCompletableFuture`  try again to rexecute the code (5 times) when index are not ready. See https://github.com/redhat-developer/lsp4ij/blob/6927714f8f2be016656eed9f4cd6d587a143f804/src/main/java/com/redhat/devtools/lsp4ij/internal/PromiseToCompletableFuture.java#L115

**Comment by turkeylurkey** [comment]
Fixed by commit `79b5c1f6eb73` "Rethrow IndexNotReadyException, ProcessCanceledException and CancellationException so the operation can be retried." in branch `useLSP4IJ1201`.
